flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -213,19 +215,23 @@ public SingleInputGate(
 	}
 
 	@Override
-	public void setup() throws IOException, InterruptedException {
+	public void setup() throws IOException {
```

Comment:
could we pass the state recovery parameters here? And have a single step `setup` method instead of introducing 2nd initialisation step (`setup` and `initializeState`).

Reply:
I agree that it has benefits to reduce steps for managing the lifecycle of input gate. But I thought of another potential concern if integrating it within `setup`. 

The previous assumption was that the output recovery should execute earlier than the input recovery in order to occupy more floating buffers from global, to get the benefit to speedup recovery process. Now the output recovery is executed by task thread during `StreamTask#beforeInvoke`. If we integrate the input recovery within `setup` process, then it would perform before output recovery to occupy more floating buffers in limited buffers environment.

I also considered integrating the output recovery within `ResultPartition#setup`, but the output recovery is executed by task thread then it would block the following operations, so I wonder it might also bring potential risks to do so.

Reply:
> The previous assumption was that the output recovery should execute earlier than the input recovery in order to occupy more floating buffers from global, to get the benefit to speedup recovery process. Now the output recovery is executed by task thread during StreamTask#beforeInvoke. If we integrate the input recovery within setup process, then it would perform before output recovery to occupy more floating buffers in limited buffers environment.

Ok, I haven't considered that. I guess it's best to keep it as it is?

Reply:
Yes, I guess so.

I found another benefit to execute the input state recovery inside `StreamTask` instead of `Task` stack(setup integration). The new dedicated executor for unspilling channel state is only meaningful for the unaligned checkpoint mode. So I tried not to create this executor for normal checkpoints, also for batch jobs to reduce total threads in TaskManager. This improvement can only be done inside `StreamTask` because it can judge the checkpoint mode via `StreamConfig`. And task class can not get this option because of `runtime` module can not depend on `streaming` module.